lastgersync/evmdownloader_fep.go

```diff
@@ -0,0 +1,210 @@
+	"github.com/agglayer/aggkit/sync"
+	aggkittypes "github.com/agglayer/aggkit/types"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+)
+
+// Event is the combination of the events that are emitted by the L2 GER manager
+type Event struct {
+	GERInfo  *GlobalExitRootInfo
+	GEREvent *GEREvent
+}
+
+type downloaderFEP struct {
+	*sync.EVMDownloaderImplementation
+	l2GERManager   *polygonzkevmglobalexitrootv2.Polygonzkevmglobalexitrootv2
+	l2GERAddr      common.Address
+	l1InfoTreeSync L1InfoTreeQuerier
+	processor      *processor
+	rh             *sync.RetryHandler
+}
+
+func newDownloaderFEP(
+	l2Client aggkittypes.BaseEthereumClienter,
+	l2GERAddr common.Address,
+	l1InfoTreeSync L1InfoTreeQuerier,
+	processor *processor,
+	rh *sync.RetryHandler,
+	blockFinality *big.Int,
+	waitForNewBlocksPeriod time.Duration,
+) (*downloaderFEP, error) {
+	l2GERManager, err := polygonzkevmglobalexitrootv2.NewPolygonzkevmglobalexitrootv2(
+		l2GERAddr, l2Client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize L2 GER manager contract: %w", err)
+	}
+
+	evmDownloader := sync.NewEVMDownloaderImplementation(
+		"lastgersync", l2Client, blockFinality,
+		waitForNewBlocksPeriod, nil, nil,
+		rh, nil)
+
+	return &downloaderFEP{
+		EVMDownloaderImplementation: evmDownloader,
+		l2GERManager:                l2GERManager,
+		l2GERAddr:                   l2GERAddr,
+		l1InfoTreeSync:              l1InfoTreeSync,
+		processor:                   processor,
+		rh:                          rh,
+	}, nil
+}
+
+// RuntimeData returns the runtime data: chainID + addresses to query
+func (d *downloaderFEP) RuntimeData(ctx context.Context) (sync.RuntimeData, error) {
+	chainID, err := d.ChainID(ctx)
+	if err != nil {
+		return sync.RuntimeData{}, err
+	}
+	return sync.RuntimeData{
+		ChainID:   chainID,
+		Addresses: []common.Address{d.l2GERAddr},
+	}, nil
+}
+
+func (d *downloaderFEP) Download(ctx context.Context, fromBlock uint64, downloadedCh chan sync.EVMBlock) {
+	var (
+		attempts            int
+		nextL1InfoTreeIndex uint32
+		err                 error
+	)
+
+	// Determine the next index to start fetching GERs
+	for {
+		latestL1InfoTreeIndex, err := d.processor.getLatestL1InfoTreeIndex()
+		if errors.Is(err, db.ErrNotFound) {
+			nextL1InfoTreeIndex = 0
+		} else if err != nil {
+			log.Errorf("error getting latest l1 info tree index: %v", err)
+			attempts++
+			d.rh.Handle("getLatestL1InfoTreeIndex", attempts)
+
+			continue
+		}
+		if latestL1InfoTreeIndex > 0 {
+			nextL1InfoTreeIndex = latestL1InfoTreeIndex + 1
+		}
+		break
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Debug("aborting the lastgersync downloader...")
+			close(downloadedCh)
+
+			return
+		default:
+		}
+
+		// Wait for new blocks before processing
+		fromBlock = d.WaitForNewBlocks(ctx, fromBlock)
+
+		// Fetch GERs from the determined index
+		attempts = 0
+		var gers []*GlobalExitRootInfo
+		for {
```

Comment:
Do you need this for loop? It looks to me, you don't need it. If `getGERsFromIndex` returns an error, just continue the main for loop.

Reply:
It was added because of retry handler, that keeps track of how many times the invocation of certain function has failed, up to predefined [limit](https://github.com/agglayer/aggkit/blob/aa160bd90b79c2074252212341cfe547d2598eb3/sync/common.go#L22-L27) (and it also [sleeps](https://github.com/agglayer/aggkit/blob/aa160bd90b79c2074252212341cfe547d2598eb3/sync/common.go#L28) for some preconfigured amount of time).

Therefore I would keep this for loop, if I'm not missing out on something.